### PR TITLE
Add Azure Video Analyzer Resource

### LINKED ISSUE #12603: Support for Azure Video Analyzer
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

[Azure Video Analyzer](https://azure.microsoft.com/en-us/products/video-analyzer/) is a new preview Azure resource, meant to be a replace for the Live Video Analytics. It's encouraged to migrate from LVA to AVA, but in order to do so for terraform setups new resources are needed.

Looking at the [Azure API endpoint](https://docs.microsoft.com/en-us/rest/api/videoanalyzer/video-analyzers/create-or-update#storageaccount) and [ARM template found sample for deploying AVA](https://docs.microsoft.com/en-us/azure/azure-video-analyzer/video-analyzer-docs/get-started-detect-motion-emit-events) and specifically the [video-analyzer.deploy.json](https://raw.githubusercontent.com/Azure/video-analyzer/main/setup/video-analyzer.deploy.json) referenced there, there are two potential terraform resources to be created from `Microsoft.Media/videoAnalyzers` and `Microsoft.Media/videoAnalyzers/edgeModules`

### New or Affected Resource(s)

* azurerm_video_analyzer
* azurerm_video_analyzer_edge_module

### Potential Terraform Configuration

Inspired by: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/media_services_account

```hcl
resource "azurerm_video_analyzer" "example" {
  name                = "example_video_analyzer"
  resource_group_name = azurerm_resource_group.example.name
  location            = local.location

  storage_account {
    id         = azurerm_storage_account.example.id
    is_primary = true # The API doesn't state how to set a storage account to be primary, only that the primary storage account must of a certain type
  }
  identity {
    type = "SystemAssigned"
  }
}

resource "azurerm_video_analyzer_edge_module" "example" {
  name                = "example_edge_module"
  resource_group_name = azurerm_resource_group.example.name
  video_analyzer_name = azurerm_video_analyzer.example.name
}
```

#### Example Argument Reference `azurerm_video_analyzer`

The following arguments are supported:

* `name` - (Required) Specifies the name of the Media Services Account. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) The name of the resource group in which to create the Media Services Account. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.

* `storage_account` - (Required) One or more `storage_account` blocks as defined below.

* `identity` - (Optional) An `identity` block is documented below.

* `storage_authentication_type` - (Optional) Specifies the storage authentication type. 
Possible value is  `ManagedIdentity` or `System`.

* `tags` - (Optional) A mapping of tags assigned to the resource.

---

A `storage_account` block supports the following:

* `id` - (Required) Specifies the ID of the Storage Account that will be associated with the Media Services instance.

* `is_primary` - (Required) Specifies whether the storage account should be the primary account or not. Defaults to `false`.

~> **NOTE:** Whilst multiple `storage_account` blocks can be specified - one of them must be set to the primary

---

A `identity` block supports the following:

* `type` - (Required) Specifies the type of Managed Service Identity that should be configured on this Media Services Account. Possible value is  `SystemAssigned`. 

---

### References

* Azure API link: <https://docs.microsoft.com/en-us/rest/api/videoanalyzer/>
* Video analyzer in the go sdk: <https://github.com/Azure/azure-sdk-for-go/tree/main/services/preview/videoanalyzer/mgmt/2021-05-01-preview/videoanalyzer>

### PR DESCRIPTION
**Pull request by EliiseS**
Add support for the Azure Video Analyzer described at #12603 

The PR does not include: 
- The account encryption parameter: https://docs.microsoft.com/en-us/rest/api/videoanalyzer/video-analyzers/create-or-update#accountencryption
- The `azurerm_video_analyzer_edge_module` resource.

> I've also finished implementing the  `azurerm_video_analyzer_edge_module`, but I'll wait for this one to be merged before I make that PR. The branch with the code is here: https://github.com/EliiseS/terraform-provider-azurerm/tree/feature/azure-video-analyzer-edge-module-resource

Fixes #12603

### CODE REVIEW

**Comment by EliiseS** [comment]
`run golangci-lint` seems to failing locally and in the pipeline due to timeout. I can't even get it to succeed locally against master

**Comment by gro1m** [comment]
I am not quite sure what you are referring to and I am not a maintainer, but as far as I understand it has to be TypeList when MaxItems equals 1 here (you cannot choose TypeSet): https://github.com/terraform-providers/terraform-provider-azurerm/pull/12665/commits/98f6efd951a51c51e684169c37041c4131ea6bcb#diff-88fdebfe75af3de1c7ca806f2126897c45721583983aa637b69d7074cc9d9ff4L60. At least this resolved the error for me when I ran into this once.

**Comment by EliiseS** [comment]
@gro1m The error has resolved itself :) I think it was a transient error

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (gro1m) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:60]: @EliiseS I am not sure because the linter error is still here?! And as far as I know it should be `pluginsdk.TypeList` when you use `MaxItems: 1`, but it is really only to the best of my knowledge...

Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:60]: All the build passed, so there's no error as far as I can tell

Comment (gro1m) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:60]: If I scroll here (https://github.com/terraform-providers/terraform-provider-azurerm/pull/12665/files) to the `video_analyzer_resource.go` file I see 
```
Check failure on line 59 in azurerm/internal/services/videoanalyzer/video_analyzer_resource.go

GitHub Actions
/ tflint
azurerm/internal/services/videoanalyzer/video_analyzer_resource.go#L59
S018: schema should use TypeList with MaxItems 1
```

but I also do not know why this is really mandatory - would need to look into S018 :)

Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:60]: Oh interesting! I'll def take a look at it tomorrow then, odd that it doesn't show up elsewhere 😱 Thanks for pointing it out!

Comment (manicminer) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:60]: I suspect this check is because there's no reason to use a set with a single item when a list has better usability

Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go]: I've gone with using a list now :)

**Comment by jackofallops** [comment]
> `run golangci-lint` seems to failing locally and in the pipeline due to timeout. I can't even get it to succeed locally against master

We're aware there's a problem with the linter / lint job, but haven't managed to run it down yet...

**Comment by EliiseS** [comment]
@jackofallops It timed out again, can you rerun the task or should I make a fake commit?

**Comment by jackofallops** [comment]
> @jackofallops It timed out again, can you rerun the task or should I make a fake commit? Or if you could review and make actual comments :)

Hi @EliiseS - The problem linter has been disabled for now on `master`, so a rebase will pick that up for you locally also. One of the team will get to the review as soon as they can, I'll mark this for review so the team get pinged though.

**Comment by EliiseS** [comment]
@jackofallops Awesome, thanks :) I've got another PR waiting for this to be merged so that's super helpful

File: azurerm/internal/services/videoanalyzer/resourceids.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/resourceids.go:3]: In ARM these are supposed to be camelCased for different words, since this is a new resource can we make this:

```suggestion
//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=VideoAnalyzer -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.Media/videoAnalyzers/analyzer1
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:93]: this would highlight a bug, in new resources we're making this case sensitive - so can we make this:

```suggestion
							ValidateFunc: validation.StringInSlice([]string{
								string("UserAssigned"),
							}, false),
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:194]: technically this is setting, we're flattening above :)

```suggestion
		return fmt.Errorf("setting `identity`: %s", err)
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:214]: this would indicate a bug (and is actually something we need to work through the rest of the provider to remove) - so can we remove this:

```suggestion
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:215]: we can make this:

```suggestion
		return fmt.Errorf("deleting %s: %+v", *id, err)
```

which'll output the same thing, using the formatting `String()` method from the ID Parser that's generated (above) to say the same thing

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:253]: since this is a nested object these fields always need to have a value (even if the API returns them as nil) - so that we can track the diff - as such can we make this:

```suggestion
		storageAccountId := ""
		if storageAccount.ID != nil {
			storageAccountId = *storageAccount.ID
		}

		userAssignedIdentityId := ""
		if storageAccount.Identity != nil && storageAccount.Identity.UserAssignedIdentity != nil {
			userAssignedIdentityId = *storageAccount.Identity.UserAssignedIdentity
		}

		results = append(results, map[string]interface{}{
			"id": storageAccountId,
			"user_assigned_identity_id": userAssignedIdentityId,
		})
```

File: website/docs/r/video_analyzer.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/video_analyzer.html.markdown:69]: this should be implied from the two role assignments, which depend on this?

```suggestion
```

File: website/docs/r/video_analyzer.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/video_analyzer.html.markdown:98]: perhaps this'd be clearer as:

```suggestion
* `user_assigned_identity_id` - (Required) Specifies the User Assigned Identity ID which should be assigned to a access this Storage Account.
```

File: website/docs/r/video_analyzer.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/video_analyzer.html.markdown:104]: `Managed Service Identity` is being referred to as a `Managed Identity` in newer Azure API's (although the term's used interchangeably), perhaps it's worth doing the same here:

```suggestion
* `type` - (Required) Specifies the type of Managed Identity that should be configured on this Video Analyzer. At this time the only possible value is `UserAssigned`. 
```

File: website/docs/r/video_analyzer.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/video_analyzer.html.markdown:106]: perhaps this'd read clearer as:

```suggestion
* `identity_ids` - (Required) A list of User Assigned Identity ID's to be assigned to this Video Analyzer.
```

File: website/docs/r/video_analyzer.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/video_analyzer.html.markdown:131]: (as above) ARM says segments should be camelCased so we should match that (since we're defining these ID's in Terraform)

```suggestion
terraform import azurerm_video_analyzer.analyzer /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/mygroup1/providers/Microsoft.Media/videoAnalyzers/analyzer1
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:71]: since this has to be a User Assigned Identity ID arguably this'd be clearer as:

```suggestion
						"user_assigned_identity_id": {
```

WDYT?

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:68]: the [storage validation package](https://github.com/terraform-providers/terraform-provider-azurerm/blob/6ac29da098b1964a65164f9ab5b0688e1776933d/azurerm/internal/services/storage/validate/storage_account_id.go#L11) contains a validation function we can use to ensure this is a Storage Account here:

```suggestion
							ValidateFunc: storageValidate.StorageAccountID,
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:80]: I maybe wrong, but I don't think this lintignore is needed?

```suggestion
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:171]: the ID formatter's `String()` method means we can reduce this to:

```suggestion
		return fmt.Errorf("retrieving %s: %+v", *id, err)
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:166]: the ID formatter's `String()` method means we can reduce this to:

```suggestion
			log.Printf("[INFO] %s was not found - removing from state", *id)
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:287]: this shouldn't be possible since both fields within this block are Required, so I think we can remove this?

```suggestion
	}
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:262]: since the `identity` block is required this shouldn't be possible, so we could raise an error here I guess?

```suggestion
		return nil, fmt.Errorf("an `identity` block is required")
```

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:274]: there is technically a potential panic here, but since both fields are required in the schema this is fine

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (tombuildsstuff) [inline azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:320]: Since this is a pointer in the SDK but needs to be a string, we'll need to conditionally pull this out incase it's nil (for `type`) - we also need to ensure the field `identity_ids` always has a value since this is a nested object:

```suggestion
	identityType := ""
	if identity.Type != nil {
		identityType = *identity.Type
	}

	identityIds := make([]interface{}, 0)
	if identity.UserAssignedIdentities != nil {
		/*
		   "userAssignedIdentities": {
		     "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/id1": {
		       "principalId": "00000000-0000-0000-0000-000000000000",
		       "clientId": "00000000-0000-0000-0000-000000000000"
		     },
		   }
		*/
		for key := range identity.UserAssignedIdentities {
			parsedId, err := msiparse.UserAssignedIdentityID(key)
			if err != nil {
				return nil, err
			}
			identityIds = append(identityIds, parsedId.ID())
		}
	}

	return []interface{}{
		map[string]interface{}{
			"type": identityType,
			"identity_ids": identityIds,
		},
	}, nil
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @EliiseS

Thanks for pushing those changes - apologies for the delayed re-review here.

I've taken a look through and left some comments inline, but this is looking good - if we can fix those up then we should be able to run the tests and get this merged 👍 

Thanks!

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:283]: You're right. There was some trial and error here, due to the fact the docs don't say it's required, but when you try the API it does. I originally built it as non-required, so there's some redundancy from that 

Commiting your suggestion :)

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:71]: All the other identity lists are called `identity_id`, which is why I went this one, for sake of consistency. If you're not planning on having other resources call this also `user_assigned_identity_id`, I wouldn't change it.

Also, aren't all `identity_id`'s provided user assigned? If not, then it could be that in the future this resource has more options for identity, which would make this a breaking change

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:68]: Oooh, thanks!

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (EliiseS) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:275]: I'll removed this if along with the below error, what do you think?

File: website/docs/r/video_analyzer.html.markdown
Comment (EliiseS) [reply website/docs/r/video_analyzer.html.markdown:69]: Great catch

File: website/docs/r/video_analyzer.html.markdown
Comment (EliiseS) [reply website/docs/r/video_analyzer.html.markdown:98]: This again, comes down to, do we want to make this change if other resources don't use this term?

**Comment by EliiseS** [comment]
@tombuildsstuff Thanks for your very thorough review! Really appreciate it, especially your insights into the newer conventions 😃 

I've resolved almost everything, except 3 issues I made additional comments on.

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (manicminer) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go:70]: More explicit is generally preferred and `user_assigned_identity_id` supports that, and is likely the convention we'll aim for across all resources - despite existing resources having to wait until the next major version :)

File: azurerm/internal/services/videoanalyzer/video_analyzer_resource.go
Comment (manicminer) [reply azurerm/internal/services/videoanalyzer/video_analyzer_resource.go]: That should be fine - if the schema changes we can add new validation accordingly 👍

File: website/docs/r/video_analyzer.html.markdown
Comment (manicminer) [reply website/docs/r/video_analyzer.html.markdown]: As per the other thread, we may as well use `user_assigned_identity_id` as that will converge with other resources when we are able to update them

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @EliiseS, I've taken a look though and this mostly LGTM. I commented on the outstanding items, could you update these and then rebase onto `main` to align with the updated directory structure (removal of top-level `azurerm/` directory)?

Once that's pushed, and the tests pass on our CI server, this should be good to merge 🚀

**Comment by EliiseS** [comment]
@manicminer I've completed the merge and changed the `identity_id` :smile:

**Review by tombuildsstuff — APPROVED** [review]
hey @EliiseS 

Thanks for rebasing this - taking a look through this now LGTM 👍 

Thanks!

**Comment by tombuildsstuff** [comment]
Tests look good 👍

**Pull request merged**
